tests/test-runner/bin/test-runner.py

```diff
@@ -408,15 +426,20 @@ def verify(self):
 
     def run(self, options):
         """
-        Create Cmd instances for the pre/post scripts. If the pre script
-        doesn't pass, skip this Test. Run the post script regardless.
+        Create Cmd instances for the pre/post/failsafe scripts. If the pre
+        script doesn't pass, skip this Test. Run the post script regardless.
+        If the Test is run, also run the failsafe script.
```

Comment:
You can also update the `tests/test-runner/man/test-runner.1` man page to describe the new failsafe options.  One thing which isn't clear to when reading through this change is if we always run the failsafe, or only when a test fails (or is killed)?

Reply:
The script will always run, but it wouldn't be hard to make it only run if the test is killed, if that seems better. I have it always run just in case there are tests that fail to clean up correctly, too. For example if there is cleanup that uses log_must either directly or indirectly through a function call, where failing could potentially skip the rest of cleanup. Ideally all the tests should not be prone to this sort of failure, but that isn't necessarily true.

Reply:
A concern with having it always run is that it currently effectively doubles the number of reported tests run.  Only having it run on a FAILED/KILLED test would be one way to largely address this.  Otherwise, we'd want to exclude these results some other way.  I don't feel strongly about it, but if I had to choose I think my preference would be to only run it on FAILED/KILLED tests.

```
Results Summary
FAIL	  13
SKIP	  28
PASS	 2558
```

Reply:
I'd agree with Brian and go a little further, suggesting that only KILLED tests should run the failsafe. It seems to me that a FAILED test that leaves the world in a bad state indicates a bug in the test that could be fixed there.

Reply:
I made it run only when a test is killed now, as that is the real problem I wanted to address. Hypothetically there can be issues with improper cleanup as well, but no real offense comes to mind.